lib/utils/FileCache.mjs

```diff
@@ -59,24 +59,59 @@ export default class FileCache {
 	hasFileChanged(absoluteFilepath) {
 		// Get file descriptor compares current metadata against cached
 		// one and stores the result to "changed" prop.w
+
+		const fileCheckSumChanged = this._useCheckSum
+			? this.hasFileCheckSumChanged(absoluteFilepath)
+			: false;
 		const descriptor = this._fileCache.getFileDescriptor(absoluteFilepath);
-		/** @type {CacheMetadata} */
-		const meta = descriptor.meta || {};
-		const changed = descriptor.changed || meta.hashOfConfig !== this._hashOfConfig;
+
+		descriptor.meta.data ??= {};
+
+		assert(isPlainObject(descriptor.meta.data));
+
+		const configChanged = descriptor.meta.data.hashOfConfig !== this._hashOfConfig;
+
+		let changed = false;
+
+		if (this._useCheckSum) {
+			changed = configChanged || fileCheckSumChanged;
+		} else {
+			changed = configChanged || Boolean(descriptor.changed);
+		}
 
 		if (!changed) {
 			debug(`Skip linting ${absoluteFilepath}. File hasn't changed.`);
 		}
 
 		// Mutate file descriptor object and store config hash to each file.
 		// Running lint with different config should invalidate the cache.
-		if (meta.hashOfConfig !== this._hashOfConfig) {
-			meta.hashOfConfig = this._hashOfConfig;
+		if (descriptor.meta.data.hashOfConfig !== this._hashOfConfig) {
+			descriptor.meta.data.hashOfConfig = this._hashOfConfig;
 		}
 
 		return changed;
 	}
 
+	/**
+	 * @param {string} absoluteFilepath
+	 * @returns {boolean}
+	 */
+	hasFileCheckSumChanged(absoluteFilepath) {
```

Comment:
[suggestion] How about making the method private if you don't want to expose it?

```suggestion
	#hasFileCheckSumChanged(absoluteFilepath) {
```